### PR TITLE
BindAddr may be "::" without "[]" (supported)

### PR DESCRIPTION
**Pull request by jovandeginste**
(alternative is to not allow `-bind ::`)

Using `::` and `[::]` is practically equivalent (https://www.ietf.org/rfc/rfc2732.txt) depending on some context.

### CODE REVIEW

**Comment by slackpad** [comment]
@sean- can you plz. look at this one wrt. to go-sockaddr?

**Comment by sean-** [comment]
`::` and `[::]` are semantically equivalent in `go-sockaddr` and should be in Consul, too, but...

Consul doesn't need to use square brackets when establishing an IPv6 default address.  We could/should make this change, but I'd suggest we align this with the other non-BC changes for 0.8.  We could/should also interpret `::` and `[::]` to their semantic values before performing any comparisons in `command/agent/agent.go`.  Doing this correctly is a larger, more invasive change than I think we want to introduce in a point release because someone could be relying on `::` as a way of binding to `IN_ADDRANY` for v6 addresses and bypassing the call to `consul.GetPublicIPv6()`.  And for the record, it is possible to `-bind=::` right now:

```
$ consul agent -dev -bind="::" -client='::1' -advertise='::1'
```

The issue is if you don't specify an advertise address, Consul dies because you can't advertise an endpoint that's the IPv6 equiv of `-advertise=0.0.0.0/32`.

I'm also pretty indifferent, but would guess given AWS's announcement re: IPv6 today, I think this is going to be a more common source of toe-stubbing in the near future.  The question is, how far do we want to go in terms of fixing this?  I think we do it right, but that involves some non-BC changes.

TL;DR: `go-sockaddr` will parse to the semantically equiv value, but Consul is currently using the string representations and there are corner cases where the behavior could be different.

**Comment by magiconair** [comment]
#2961 bans the use of an INADDR_ANY for the advertise address. Since we support "[::]" we should support "::" if they are equivalent. I can make this change once #3028 is merged. @sean- any objections?

**Comment by sean-** [comment]
@magiconair None, go for it.
